Documentation/Setup/TsconfigReference.rst
.. include:: /Includes.rst.txt

.. _tsconfigRef:

==================
TSconfig Reference
==================

Default configuration:
https://github.com/cyberelk/brofix/blob/main/Configuration/TsConfig/Page/pagetsconfig.tsconfig

.. contents::
   :local:
   :depth: 1

.. _tsconfigUserAgent:

User-Agent
==========

*required*

.. container:: table-row

   Property
      mod.brofix.linktypesConfig.external.headers.User-Agent

   Data type
      string

   Description
      This is what is sent as "User-Agent" to the external site when
      checking external URLs. It should contain a working URL with
      contact information.

      .. code-block:: none

         User-Agent = Mozilla/5.0 (compatible; Mysite LinkChecker/1.1; +https://mysite.com/imprint.html

   Default
      If not set, a default is automatically generated using the email address from Global Configuration
      :php:`$GLOBALS['TYPO3_CONF_VARS']['MAIL']['defaultMailFromAddress']`.

.. _tsconfigLinktypes:

linktypes
=========

*optional*

.. container:: table-row

   Property
      mod.brofix.linktypes

   Data type
      string

   Description
      Comma separated list of hooks to load.

      **Possible values:**

      **db**: Check links to database records (pages, content elements).

      **file**: Check links to files located in your local TYPO3 installation.

      **external**: Check links to external files.

      This list may be extended by other extensions providing a linktype
      checker.

   Default
      db,file,external

.. _tsconfigSearchfields:

searchFields.[table]
====================

*optional*

.. container:: table-row

   Property
      mod.brofix.searchFields.[table]

   Data type
      string

   Description
      Comma separated list of table fields in which to check for
      broken links. Broken Link Fixer only checks fields that have
      been defined in :ts:`searchFields`.

      Broken Link Fixer ships with sensible defaults that work well
      for the TYPO3 core. Not all fields which contain links are
      currently checked though. You can configure additional fields
      for extensions.

      .. warning::

         Currently, Broken Link Fixer can only detect links in specific types
         of fields, as configured in TCA:

         *  fields with at least one :ref:`softref <t3tca:columns-input-properties-softref>` set
            in their TCA configuration.

         *  fields with type "link" (like tt_content.header_link)

         For this reason, it is currently not possible to check for
         `pages.media`.

         Examples for working fields:

         *  `pages.canonical_link`
         *  `pages.url`

         Examples for not working fields:

         *  `pages.media`


   Example
      Only check for `bodytext` in `tt_content`:

      .. code-block:: typoscript

         mod.brofix.searchFields.tt_content = bodytext

      Add checks for news and calendarize events:

      .. code-block:: typoscript

         mod.brofix.searchFields.tx_news_domain_model_news = bodytext
         mod.brofix.searchFields.tx_calendarize_domain_model_event = description

   Default
      .. code-block:: typoscript

         pages = media,url
         tt_content = bodytext,header_link,records


.. _tsconfigexcludeCtype:

excludeCtype
============

*optional*

.. container:: table-row

   Property
      mod.brofix.excludeCtype

   Data type
      string

   Description
      Exclude specific content types from link checking. 'html' is not
      checked by default, because the parsing for links does not always
      work correctly and may cause a number of links to be displayed as
      broken, which are in fact ok (false positives).

   Default
      html


.. _tsconfigDoNotCheckContentOnPagesDoktypes:

doNotCheckContentOnPagesDoktypes
================================

*optional*

.. container:: table-row

   Property
      mod.brofix.check.doNotCheckContentOnPagesDoktypes

   Data type
      string

   Description
      Comma separated list of page types on which content should not be
      checked. This still means the pages will get checked.

      This is for example by default the cause for the page types shortcut
      and external link.

   Default
      3,4 (Link to external URL, shortcut)


.. _tsconfigDoNotCheckPagesDoktypes:

doNotCheckPagesDoktypes
=======================

*optional*

.. container:: table-row

   Property
      mod.brofix.check.doNotCheckPagesDoktypes

   Data type
      string

   Description
      Comma separated list of page types which should not be checked.
      This means if a page has a doktype which is listed in this list,
      we do not do any link checking on the page.

   Default
      6,7,199,255 (Backend User section, Mount Point, Menu Separator, Recycler)


.. _tsconfigDoNotTraversePagesDoktypes:

doNotTraversePagesDoktypes
--------------------------

*optional*

.. container:: table-row

   Property
      mod.brofix.check.doNotTraversePagesDoktypes

   Data type
      string

   Description
      Comma separated list of page types which should not be traversed.
      This means if a page has a doktype which is listed in this list,
      we do not do any link checking on subpages of these pages (and
      subpages of the subpages etc.).

   Default
      6,199,255 (Backend User section, Menu Separator, Recycler)

.. _tsconfigDoNotCheckLinksOnWorkspace:

doNotCheckLinksOnWorkspace
==========================

*optional*

.. container:: table-row

   Property
      mod.brofix.check.doNotCheckLinksOnWorkspace

   Data type
      int

   Description
      This option is used to enable or disable checking links that are created in Workspace, by default,
      the links created on workspaces will be checked and reported.

   Default
      0


.. _tsconfigReportHiddenRecords:

reportHiddenRecords
===================

*optional*

.. container:: table-row

   Property
      mod.brofix.reportHiddenRecords

   Data type
      int

   Description
      Whether links to hidden records should be treated as broken links.

      .. important::

         This used to be `linkhandler.reportHiddenRecords` but is now available
         as configuration option for any linktype.

   Default
      1


.. _tsconfigDepth:

depth
-----

*optional*

.. container:: table-row

   Property
      mod.brofix.depth

   Data type
      int

   Description
      Default depth when checking with console command

   Default
      999 (for infinite)


Accept
======

*optional*

.. container:: table-row

   Property
      mod.brofix.linktypesConfig.external.headers.Accept

   Data type
      string

   Description
      HTTP request header "Accept". It is recommended to leave the default value and not change this.

   Default
      `*/*`

Accept-Language
===============

*optional*

.. container:: table-row

   Property
      mod.brofix.linktypesConfig.external.headers.Accept

   Data type
      string

   Description
      HTTP request header. It is recommended to leave the default value and not change this.

   Default
      `*`


Accept-Encoding
===============

*optional*

.. container:: table-row

   Property
      mod.brofix.linktypesConfig.external.headers.Accept-Encoding

   Data type
      string

   Description
      HTTP request header. It is recommended to leave the default value and not change this.

   Default
      `*`

timeout
=======

*optional*

.. container:: table-row

   Property
      mod.brofix.linktypesConfig.external.timeout

   Data type
      int

   Description
      Timeout for HTTP request.

   Default
      10

redirects
=========

*optional*

.. container:: table-row

   Property
      mod.brofix.linktypesConfig.external.redirects

   Data type
      int

   Description
      Number of redirects to follow. If more redirects are necessary to reach
      the destination final URL, this is handled as broken link.

   Default
      5

.. _tsConfigExcludeLinkTargetStoragePid:

excludeLinkTarget.storagePid
============================

*required* (if "exclude URL" functionality should be available for non-admin
editors)

.. container:: table-row

   Property
      mod.brofix.excludeLinkTarget.storagePid

   Data type
      int

   Description
      The pid of the storage folder which contains the excluded link target
      records. If you want to enable editors to add URLs to list of excluded
      URLs, you must change this (it must be != 0).

      Create a central folder to store the excluded URLs or create one for each
      site.

      .. important::

         The storage pid is stored along with the broken link records. If
         you change this value, you should start a complete recheck of broken
         links to get this updated.

      Excluded link targets (=URLs) are treated as valid URLs. This can be
      used for the **rare** case that an URL is detected as broken, but is
      not broken. This may be the case for some sites which require login
      credentials, but also for common sites where the automatic link
      checking mechanism yields false results.

   Default
      0


excludeLinkTarget.allowed
=========================

*optional*

.. container:: table-row

   Property
      mod.brofix.excludeLinkTarget.allowed

   Data type
      string

   Description
      Allowed link types which can be excluded. By default, it is only possible
      to exclude external URLs. If you would like to make this available for
      page links too, add additional link types, e.g.

      .. code-block:: typoscript

         allowed = external,db

      You can set it to empty to disable the "exclude URL" functionality:

      .. code-block:: typoscript

         allowed =

   Default
      external


.. _tsconfigLinkTargetCacheExpiresLow:

linkTargetCache.expiresLow
==========================

*optional*

.. container:: table-row

   Property
      mod.brofix.linkTargetCache.expiresLow

   Data type
      int

   Description
      When the link target cache expires in seconds. Whenever an external URL
      is checked or rechecked, the link target cache is used. Once the cache
      expires, the URL must be checked again.

      The value means that the information for external URLs is retained for
      that time without having to access the external site.

      2 different values are used for expiresLow and expiresHigh so that the
      target will usually not expire during the on-the-fly checking which would
      lead to delays.

      As a rule of thumb, use the interval for full checking (e.g. 1 day for
      once a day checking) and multiply that with a factor of 1 to 10 for
      expiresLow. Add another interval for expiresHigh.

      The interval for expiresLow will be used for full checking via the
      console command.

      .. code-block:: typoscript

         # checking links daily, use 7 as factor:
         #  1 day * 7 * (seconds per day)
         #  1 * 7 * 24*60*60
         linkTargetCache.expiresLow = 604800
         #  1 * 8 * 24*60*60
         linkTargetCache.expiresHigh = 691200

   Default
      604800 (7 days)

.. _tsconfigLinkTargetCacheExpiresHigh:

linkTargetCache.expiresHigh
===========================

*optional*

.. container:: table-row

   Property
      mod.brofix.linkTargetCache.expiresHigh

   Data type
      int

   Description
      See :ref:`tsconfigLinkTargetCacheExpires` for description

   Default
      691200 (8 days)

.. _tsconfig_crawlDelay_seconds:

crawlDelay.seconds
==================

*optional*

.. container:: table-row

   Property
      mod.brofix.crawlDelay.seconds

   Data type
      int

   Description
      The **minimum** number of seconds that must have passed between
      checking 2 URL for the same domain.

      If the required time has already passed since an URL of the same domain
      was last checked, the wait is not performed.

      This helps to prevent that external sites are bombarded with requests from
      our site.

      .. note::

         Currently, a wait is not performed for every URL if URLs are redirected
         because this is handled internally by Guzzle.

      This is a pragmatic approach to make sure that a minimum delay is used
      when checking URLs of the same site. As a site may have multiple domains
      or several domains may be used by the same site, this will not always get
      the desired result, but it is a "good enough" approach.

      This will not be used for :ref:`on-the-fly <linkCheckingOnTheFly>`
      checking, only for checking via the console command task.

      .. code-block:: typoscript

         crawlDelay.seconds = 10

   Default
      5


crawlDelay.nodelay
==================

*optional*

.. container:: table-row

   Property
      mod.brofix.crawlDelay.nodelay

   Data type
      string

   Description
      Do not use the crawlDelay.seconds wait period for these domains


      .. code-block:: typoscript

         crawlDelay.nodelay = example.org,example.com

      .. code-block:: typoscript
         :caption: regex example
         crawlDelay.nodelay = regex:/(.*\.)?example.(org|com)/

   Default
      empty


report.docsurl
==============

*optional*

.. container:: table-row

   Property
      mod.brofix.report.docsurl

   Data type
      string

   Description
      Add a documentation URL. This will add an "i" button to the broken link
      report with a link to the documentation.


Classes/Configuration/Configuration.php
<?php

declare(strict_types=1);

namespace Cyberelk\Brofix\Configuration;

use Symfony\Component\Mime\Address;
use Cyberelk\Brofix\FormEngine\FieldShouldBeChecked;
use Cyberelk\Brofix\FormEngine\FieldShouldBeCheckedFull;
use Cyberelk\Brofix\FormEngine\FieldShouldBeCheckedWithFlexform;
use Cyberelk\Brofix\Linktype\AbstractLinktype;
use Cyberelk\Brofix\Linktype\LinktypeInterface;
use TYPO3\CMS\Backend\Utility\BackendUtility;
use TYPO3\CMS\Core\Utility\ArrayUtility;
use TYPO3\CMS\Core\Utility\ExtensionManagementUtility;
use TYPO3\CMS\Core\Utility\GeneralUtility;
use TYPO3\CMS\Core\Utility\MailUtility;

/**
 * Wrapper class for handling TSconfig. Using the convenience
 * functions in this class the properties can be accessed
 * directly and the functions will already return the correct type.
 *
 * Configuration can still be extended by extending the TSconfig mod.brofix
 * and the function getTsConfig() used in this class.
 */
class Configuration
{
    public const SEND_EMAIL_NEVER = 'never';

    /** @var string Always send email */
    public const SEND_EMAIL_ALWAYS = 'always';

    /** @var string Send email if any broken links found */
    public const SEND_EMAIL_ANY = 'any';

    /** @var string Send email if any NEW broken links found */
    public const SEND_EMAIL_NEW = 'new';

    /** @var string Do not override configuration for email sending */
    public const SEND_EMAIL_AUTO = 'auto';

    public const SEND_EMAIL_DEFAULT_VALUE = self::SEND_EMAIL_AUTO;

    public const SEND_EMAIL_AVAILABLE_VALUES = [self::SEND_EMAIL_NEVER, self::SEND_EMAIL_ALWAYS, self::SEND_EMAIL_ANY,
        self::SEND_EMAIL_NEW, self::SEND_EMAIL_AUTO];

    public const SHOW_EDIT_BUTTONS_EDIT_FIELD = 'field';
    public const SHOW_EDIT_BUTTONS_EDIT_FULL = 'full';

    public const SHOW_EDIT_BUTTONS_BOTH = 'both';

    public const SHOW_EDIT_BUTTONS_DEFAULT_VALUE = self::SHOW_EDIT_BUTTONS_BOTH;

    public const TRAVERSE_MAX_NUMBER_OF_PAGES_IN_BACKEND_DEFAULT = 1000;
    public const DEFAULT_TSCONFIG = [
        'searchFields.' => [
            'pages' => 'media,url',
            'tt_content' => 'bodytext,header_link,records',
        ],
        'excludeCtype' => 'html',
        'linktypes' => 'db,file,external,applewebdata',
        'check.' => [
            'doNotCheckContentOnPagesDoktypes' => '3,4',
            'doNotCheckPagesDoktypes' => '6,7,199,255',
            'doNotTraversePagesDoktypes' => '6,199,255',
            'doNotCheckLinksOnWorkspace' => false,
        ],
        'checkhidden' => false,
        'depth' => 999,
        'reportHiddenRecords' => true,
        'linktypesConfig.' => [
            'external.' => [
                'headers.' => [
                    'User-Agent' => '',
                    'Accept' => '*/*'
                ],
                'timeout' => 10,
                'redirects' => 5,
            ]
        ],
        'excludeLinkTarget.' => [
            'storagePid' => 0,
            'allowed' => 'external',
        ],
        'linkTargetCache.' => [
            'expiresLow' => 604800,
            'expiresHigh' => 691200,
        ],
        'crawlDelay.' => [
            'seconds' => 5,
            'nodelay' => '',
        ],
        'report.' => [
            'docsurl' => '',
            'recheckButton' => -1,
        ],
        'mail.' => [
            'sendOnCheckLinks' => 1,
            'recipients' => '',
            'fromname' => '',
            'fromemail' => '',
            'replytoname' => '',
            'replytoemail' => '',
            'subject' => '',
            'template' => 'CheckLinksResults',
            'language' => 'en',
        ],
        'custom.' => []
    ];

    protected const TCA_PROCESSING_DEFAULT_VALUE = self::TCA_PROCESSING_DEFAULT;

    public const TCA_PROCESSING_NONE = 'none';
    public const TCA_PROCESSING_DEFAULT = 'default';
    public const TCA_PROCESSING_FULL = 'full';

    /**
     * @var mixed[]
     * @todo replace this array and use member variables for all values
     */
    protected $tsConfig = self::DEFAULT_TSCONFIG;

    /**
     * Limit number of pages traversed in backend. This limit is only active when displaying
     * and checking links in the Backend, not when checking links via CLI or scheduler
     *
     * We set to 0 here by default (which means no limit) to not limit link checking via CLI.
     *
     * @var int
     */
    protected int $traverseMaxNumberOfPagesInBackend = 0;

    /** Show all links, not just the broken links */
    protected bool $showAllLinks = true;

    protected string $combinedErrorNonCheckableMatch = 'regex:/^httpStatusCode:(401|403):/';

    /**
     * @var string[]
     *
     * Do not use these softreference parsers (only in fields $excludeSoftrefsInFields)
     */
    protected $excludeSoftrefs;

    /** @var string[] */
    protected $excludeSoftrefsInFields;

    /**
     * Array for hooks for own checks
     *
     * @var array<string,LinktypeInterface>
     */
    protected array $hookObjectsArr = [];

    protected string $tcaProcessing = self::TCA_PROCESSING_DEFAULT_VALUE;

    protected string $overrideFormDataGroup = '';

    protected string $showEditButtons = self::SEND_EMAIL_DEFAULT_VALUE;

    protected bool $showPageLayoutButton = true;

    /**
     * Configuration constructor.
     * @param array<mixed> $extConfArray ExtensionConfiguration array
     */
    public function __construct(array $extConfArray)
    {
        // initialize from extension configuration
        $this->showAllLinks = (bool)($extConfArray['showalllinks'] ?? true);
        $this->showEditButtons = ($extConfArray['showEditButtons'] ?? self::SHOW_EDIT_BUTTONS_DEFAULT_VALUE);
        $this->showPageLayoutButton = (bool)($extConfArray['showPageLayoutButton'] ?? true);
        $this->combinedErrorNonCheckableMatch = $extConfArray['combinedErrorNonCheckableMatch'] ?? '';
        $this->excludeSoftrefs = explode(',', $extConfArray['excludeSoftrefs'] ?? '');
        $this->excludeSoftrefsInFields = explode(',', $extConfArray['excludeSoftrefsInFields'] ?? '');
        $this->setTraverseMaxNumberOfPagesInBackend(
            (int)($extConfArray['traverseMaxNumberOfPagesInBackend']
                ?? self::TRAVERSE_MAX_NUMBER_OF_PAGES_IN_BACKEND_DEFAULT)
        );
        $this->tcaProcessing = $extConfArray['tcaProcessing'] ?? self::TCA_PROCESSING_DEFAULT_VALUE;
        $this->overrideFormDataGroup = $extConfArray['overrideFormDataGroup'] ?? '';

        // initialize from global configuration
        // Hook to handle own checks
        foreach ($GLOBALS['TYPO3_CONF_VARS']['EXTCONF']['brofix']['checkLinks'] ?? [] as $key => $className) {
            /**
             * @var LinktypeInterface $linktype
             */
            $linktype = GeneralUtility::makeInstance($className); // @phpstan-ignore-line
            $this->hookObjectsArr[$key] = $linktype;
            $linktype->setConfiguration($this);
        }
    }

    /**
     * @param mixed[] $tsConfig
     */
    public function setTsConfig(array $tsConfig): void
    {
        $this->tsConfig = $tsConfig;
    }

    /**
     * Get the brofix modTSconfig for a page
     *
     * @param int $page Uid of the page
     * @throws \Exception
     */
    public function loadPageTsConfig(int $page): void
    {
        $this->setTsConfig(BackendUtility::getPagesTSconfig($page)['mod.']['brofix.'] ?? []);
    }

    /**
     * @param array<mixed> $override
     */
    public function overrideTsConfigByArray(array $override): void
    {
        ArrayUtility::mergeRecursiveWithOverrule($this->tsConfig, $override);
    }

    /**
     * @return mixed[]
     */
    public function getTsConfig(): array
    {
        return $this->tsConfig;
    }

    /**
     * @param array<string,array<string>> $searchFields , e.g.
     *   [
     *      'tt_content' => ['bodytext', 'media']
     *   ]
     */
    public function setSearchFields(array $searchFields): void
    {
        unset($this->tsConfig['searchFields.']);
        foreach ($searchFields as $table => $fields) {
            $this->tsConfig['searchFields.'][$table] = implode(',', $fields);
        }
    }

    /**
     * Get the list of fields to parse in modTSconfig
     *
     * @return array<string,array<string>> $searchFields List of fields, e.g.
     *   [
     *      'tt_content' => ['bodytext', 'media']
     *   ]
     */
    public function getSearchFields(): array
    {
        if (!isset($this->tsConfig['searchFields.'])) {
            return [];
        }
        // Get the searchFields from TypoScript
        foreach ($this->tsConfig['searchFields.'] as $table => $fieldList) {
            $fields = GeneralUtility::trimExplode(',', $fieldList);
            foreach ($fields as $field) {
                $searchFields[$table][] = $field;
            }
        }
        return $searchFields ?? [];
    }

    /**
     * @return array<string>
     */
    public function getExcludedCtypes(): array
    {
        return explode(',', (string)($this->tsConfig['excludeCtype'] ?? ''));
    }

    /**
     * @param array<int,string> $linkTypes
     */
    public function setLinkTypes(array $linkTypes): void
    {
        $this->tsConfig['linktypes'] = implode(',', $linkTypes);
    }

    /**
     * @return array<string>
     */
    public function getLinkTypes(): array
    {
        return explode(',', $this->tsConfig['linktypes'] ?? 'external,db,file');
    }

    /**
     * @return array<int,int>
     */
    public function getDoNotCheckContentOnPagesDoktypes(): array
    {
        $doktypes = explode(',', $this->tsConfig['check.']['doNotCheckContentOnPagesDoktypes'] ?? '3,4');
        $result = [];
        foreach ($doktypes as $doktype) {
            $doktype = (int)$doktype;
            $result[$doktype] = $doktype;
        }
        return $result;
    }

    /**
     * @return array<int,int>
     */
    public function getDoNotCheckPagesDoktypes(): array
    {
        $doktypes = explode(',', $this->tsConfig['check.']['doNotCheckPagesDoktypes'] ?? '6,7,199,255');
        $result = [];
        foreach ($doktypes as $doktype) {
            $doktype = (int)$doktype;
            $result[$doktype] = $doktype;
        }
        return $result;
    }

    /**
     * @return array<int,int>
     */
    public function getDoNotTraversePagesDoktypes(): array
    {
        $doktypes = explode(',', $this->tsConfig['check.']['doNotTraversePagesDoktypes'] ?? '6,199,255');
        $result = [];
        foreach ($doktypes as $doktype) {
            $doktype = (int)$doktype;
            $result[$doktype] = $doktype;
        }
        return $result;
    }

    /**
     * @return bool
     */
    public function getDoNotCheckLinksOnWorkspace(): bool
    {
        return $this->tsConfig['check.']['doNotCheckLinksOnWorkspace'] == 1;
    }

    /**
     * @return bool
     */
    public function isCheckHidden(): bool
    {
        return (bool)($this->tsConfig['checkhidden'] ?? false);
    }

    public function isReportHiddenRecords(): bool
    {
        return (bool)($this->tsConfig['reportHiddenRecords'] ?? true);
    }

    /**
     * @param string $linktype
     * @return mixed[]
     */
    public function getLinktypesConfig(string $linktype): array
    {
        return $this->tsConfig['linktypesConfig.'][$linktype . '.'] ?? [];
    }

    /**
     * @return mixed[]
     */
    public function getLinktypesConfigExternalHeaders(): array
    {
        $headers = $this->tsConfig['linktypesConfig.']['external.']['headers.'] ?? [];
        if (($headers['User-Agent'] ?? '') === '') {
            $headers['User-Agent'] = $this->getUserAgent();
        }
        return $headers;
    }

    public function getUserAgent(): string
    {
        return $this->tsConfig['linktypesConfig.']['external.']['headers.']['User-Agent'] ??
            'Mozilla/5.0 (compatible; using Brofix link checker/ ' . (MailUtility::getSystemFromAddress() ?: '');
    }

    public function getLinktypesConfigExternalTimeout(): int
    {
        return (int)($this->tsConfig['linktypesConfig.']['external.']['timeout'] ?? 10);
    }

    public function getLinktypesConfigExternalRedirects(): int
    {
        return (int)($this->tsConfig['linktypesConfig.']['external.']['redirects'] ?? 5);
    }

    public function getExcludeLinkTargetStoragePid(): int
    {
        return (int)($this->tsConfig['excludeLinkTarget.']['storagePid'] ?? 0);
    }

    /**
     * @return array<string>
     */
    public function getExcludeLinkTargetAllowedTypes(): array
    {
        return explode(',', $this->tsConfig['excludeLinkTarget.']['allowed'] ?? 'external');
    }

    public function getLinkTargetCacheExpires(int $flags = 0): int
    {
        if ($flags | AbstractLinktype::CHECK_LINK_FLAG_SYNCHRONOUS) {
            return (int)($this->tsConfig['linkTargetCache.']['expiresHigh'] ?? 604800);
        }
        return (int)($this->tsConfig['linkTargetCache.']['expiresLow'] ?? 518400);
    }

    /**
     * @return int
     *
     * @deprecated Use getCrawlDelaySeconds. Accuracy is only by seconds.
     */
    public function getCrawlDelayMs(): int
    {
        trigger_error(
            'Method "getCrawlDelayMs" is deprecated since brofix version 2.0.0, should use getCrawlDelaySeconds',
            E_USER_DEPRECATED
        );

        if (isset($this->tsConfig['crawlDelay.']['seconds'])) {
            return (int)($this->tsConfig['crawlDelay.']['seconds'] * 1000);
        }
        if (isset($this->tsConfig['crawlDelay.']['ms'])) {
            return (int)($this->tsConfig['crawlDelay.']['ms']);
        }
        return 5000;
    }

    /**
     * @return int
     */
    public function getCrawlDelaySeconds(): int
    {
        return (int)($this->tsConfig['crawlDelay.']['seconds'] ?? 5);
    }

    /**
     * Return crawlDelay nodelay domains.
     * @return array<string>
     * @deprecated Use getCrawlDelayNodelayDomains
     */
    public function getCrawlDelayNodelay(): array
    {
        $noDelayString = $this->getCrawlDelayNodelayString();
        if (str_starts_with($noDelayString, 'regex:')) {
            return [];
        }
        return explode(',', $$noDelayString);
    }

    /**
     * Return crawlDelay.nodelay domains.
     * @return array<string>awlDelayNodelayDomains
     */
    public function getCrawlDelayNodelayDomains(): array
    {
        $noDelayString = $this->getCrawlDelayNodelayString();
        if (str_starts_with($noDelayString, 'regex:')) {
            return [];
        }
        return explode(',', $noDelayString);
    }

    public function isCrawlDelayNoDelayRegex(): bool
    {
        return str_starts_with($this->getCrawlDelayNodelayString(), 'regex:');
    }

    public function getCrawlDelayNoDelayRegex(): string
    {
        if ($this->isCrawlDelayNoDelayRegex()) {
            return trim(substr($this->getCrawlDelayNodelayString(), strlen('regex:')));
        }
        return '';
    }

    public function getCrawlDelayNodelayString(): string
    {
        return trim($this->tsConfig['crawlDelay.']['nodelay'] ?? '');
    }

    public function getDocsUrl(): string
    {
        return $this->tsConfig['report.']['docsurl'] ?? '';
    }

    public function getRecheckButton(): int
    {
        return (int)($this->tsConfig['report.']['recheckButton'] ?? -1);
    }

    public function getMailSendOnCheckLinks(): string
    {
        $value = ($this->tsConfig['mail.']['sendOnCheckLinks'] ?? self::SEND_EMAIL_DEFAULT_VALUE);
        switch ($value) {
            case self::SEND_EMAIL_AUTO:
                return self::SEND_EMAIL_DEFAULT_VALUE;
                // map old values
            case '0':
                return self::SEND_EMAIL_NEVER;
            case '1':
                return self::SEND_EMAIL_ALWAYS;
            case '-1':
                return self::SEND_EMAIL_DEFAULT_VALUE;
        }
        return $value;
    }

    public function setMailSendOnCheckLinks(string $value): void
    {
        if ($value !== self::SEND_EMAIL_AUTO) {
            $this->tsConfig['mail.']['sendOnCheckLinks'] = $value;
        }
    }

    /**
     * Get check depth (number of page recursion levels)
     *
     * @return int
     */
    public function getDepth(): int
    {
        return (int)($this->tsConfig['depth'] ?? 999);
    }

    public function setDepth(int $depth): void
    {
        $this->tsConfig['depth'] = $depth;
    }

    public function setMailRecipientsAsString(string $recipients): void
    {
        $this->tsConfig['mail.']['recipients'] = $recipients;
    }

    /**
     * @return Address[]
     */
    public function getMailRecipients(): array
    {
        $result = [];
        $recipients = trim($this->tsConfig['mail.']['recipients'] ?? '');
        if ($recipients === '') {
            if ($GLOBALS['TYPO3_CONF_VARS']['MAIL']['defaultMailFromAddress'] ?? '') {
                $fromAddress = $GLOBALS['TYPO3_CONF_VARS']['MAIL']['defaultMailFromAddress'] ?? '';
                $fromName = $GLOBALS['TYPO3_CONF_VARS']['MAIL']['defaultMailFromName'] ?? '';
                $result[] = new Address($fromAddress, $fromName);
            }
        } else {
            foreach (explode(',', $recipients) as $recipient) {
                $result[] = Address::create($recipient);
            }
        }
        return $result;
    }

    public function getMailTemplate(): string
    {
        return $this->tsConfig['mail.']['template'] ?? 'CheckLinksResults';
    }

    /**
     * From email address. Returns only the email address.
     *
     * @return string
     */
    public function getMailFromEmail(): string
    {
        if (($this->tsConfig['mail.']['fromemail'] ?? '') !== '') {
            return $this->tsConfig['mail.']['fromemail'];
        }
        if (($this->tsConfig['mail.']['from'] ?? '') !== '') {
            // this is deprecated
            return explode(' ', $this->tsConfig['mail.']['from'])[0];
        }
        return MailUtility::getSystemFromAddress() ?: '';
    }

    /**
     * From Name
     *
     * @return string
     */
    public function getMailFromName(): string
    {
        if (($this->tsConfig['mail.']['fromname'] ?? '') !== '') {
            return $this->tsConfig['mail.']['fromname'];
        }
        return MailUtility::getSystemFromName() ?: '';
    }

    public function getMailReplyToEmail(): string
    {
        if (($this->tsConfig['mail.']['replytoemail'] ?? '') !== '') {
            return $this->tsConfig['mail.']['replytoemail'];
        }
        if (($this->tsConfig['mail.']['replyto'] ?? '') !== '') {
            // this is deprecated
            return explode(' ', $this->tsConfig['mail.']['replyto'])[0];
        }
        // @todo use MailUtility::getSystemReplyTo()
        // getSystemReplyto returns an array of email addresses, can be in format
        // 'name <email@from>'
        return '';
    }

    /**
     * From Name
     *
     * @return string
     */
    public function getMailReplyToName(): string
    {
        if (($this->tsConfig['mail.']['replytoname'] ?? '') !== '') {
            return $this->tsConfig['mail.']['replytoname'];
        }
        // @todo use MailUtility::getSystemReplyTo()
        // getSystemReplyto returns an array of email addresses, can be in format
        // 'name <email@from>'
        return '';
    }

    public function getMailSubject(): string
    {
        return $this->tsConfig['mail.']['subject'] ?? '';
    }

    public function getMailLanguage(): string
    {
        return $this->tsConfig['mail.']['language'] ?? 'en';
    }

    /**
     * @return int
     */
    public function getTraverseMaxNumberOfPagesInBackend(): int
    {
        return $this->traverseMaxNumberOfPagesInBackend;
    }

    /**
     * @param int $traverseMaxNumberOfPagesInBackend
     */
    public function setTraverseMaxNumberOfPagesInBackend(int $traverseMaxNumberOfPagesInBackend): void
    {
        $this->traverseMaxNumberOfPagesInBackend = $traverseMaxNumberOfPagesInBackend;
    }

    /**
     * @return string[]
     */
    public function getExcludeSoftrefs(): array
    {
        return $this->excludeSoftrefs;
    }

    /**
     * @return string[]
     */
    public function getExcludeSoftrefsInFields(): array
    {
        return $this->excludeSoftrefsInFields;
    }

    public function getShowEditButtons(): string
    {
        return $this->showEditButtons;
    }

    public function setShowEditButtons(string $showEditButtons): void
    {
        $this->showEditButtons = $showEditButtons;
    }

    public function isShowPageLayoutButton(): bool
    {
        // check if page_layouts is loaded
        if (!ExtensionManagementUtility::isLoaded('page_callouts')) {
            return false;
        }

        return $this->showPageLayoutButton;
    }

    public function setShowPageLayoutButton(bool $showPageLayoutButton): void
    {
        $this->showPageLayoutButton = $showPageLayoutButton;
    }

    public function isShowAllLinks(): bool
    {
        return $this->showAllLinks;
    }

    public function setShowAllLinks(bool $showAllLinks): void
    {
        $this->showAllLinks = $showAllLinks;
    }

    public function getCombinedErrorNonCheckableMatch(): string
    {
        return $this->combinedErrorNonCheckableMatch;
    }

    public function getTcaProcessing(): string
    {
        return $this->tcaProcessing;
    }

    public function getOverrideFormDataGroup(): string
    {
        return $this->overrideFormDataGroup;
    }

    public function getFormDataGroup(): string
    {
        if ($this->overrideFormDataGroup) {
            return $this->overrideFormDataGroup;
        }
        switch ($this->getTcaProcessing()) {
            case self::TCA_PROCESSING_NONE:
                return '';
            case self::TCA_PROCESSING_DEFAULT:
                return FieldShouldBeChecked::class;
            case self::TCA_PROCESSING_FULL:
                //return FieldShouldBeCheckedFull::class;
                return FieldShouldBeCheckedWithFlexform::class;
        }
        return '';
    }

    public function getLinktypeObject(string $linktype): ?LinktypeInterface
    {
        return $this->hookObjectsArr[$linktype] ?? null;
    }

    /**
     * @return array<string,LinktypeInterface>
     */
    public function getLinktypeObjects(): array
    {
        return $this->hookObjectsArr;
    }

    /**
     * Get custom configuration (user defined)
     * @return mixed[]
     */
    public function getCustom(): array
    {
        return $this->tsConfig['custom.'] ?? [];
    }
}
